source/dofs/dof_tools.cc

```diff
@@ -765,6 +765,140 @@ namespace DoFTools
+    local_dof_indices.reserve (max_dofs_per_cell(dof_handler));
+
+    // Get all the dofs that live on the subdomain:
+    std::set<types::global_dof_index> predicate_dofs;
+
+    typename DoFHandlerType::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (predicate_local(cell))
+        {
+          local_dof_indices.resize (cell->get_fe().dofs_per_cell);
+          cell->get_dof_indices (local_dof_indices);
+          predicate_dofs.insert(local_dof_indices.begin(),
+                                local_dof_indices.end());
+        }
+
+    // Get halo layer and accumulate its DoFs
+    std::set<types::global_dof_index> halo_dofs;
+
+    const std::vector<typename DoFHandlerType::active_cell_iterator> halo_cells =
+      GridTools::compute_active_cell_halo_layer(dof_handler, predicate_local);
+    for (typename std::vector<typename DoFHandlerType::active_cell_iterator>::const_iterator it = halo_cells.begin();
+         it != halo_cells.end(); ++it)
+      {
+        // skip halo cells that satisfy the given predicate and thereby
+        // shall be a part of the index set on another MPI core.
+        // Those could only be ghost cells with p::d::Tria.
+        if (predicate(*it))
+          {
+            Assert ((*it)->is_ghost(),
+                    ExcInternalError());
+            continue;
+          }
+
+        const unsigned int dofs_per_cell = (*it)->get_fe().dofs_per_cell;
+        local_dof_indices.resize (dofs_per_cell);
+        (*it)->get_dof_indices (local_dof_indices);
+        halo_dofs.insert(local_dof_indices.begin(),
+                         local_dof_indices.end());
+      }
+
+    // A complication coming from the fact that DoFs living on locally
+    // owned cells which satisfy predicate may actually be constrained to DoFs
+    // with the support outside of the predicate. And the other way around -
+    // such DoFs may be muster DoFs for other DoFs outside of the predicate.
+    if (cm.n_constraints() > 0)
+      {
+        const std::vector<std::pair<types::global_dof_index,double> > *line_ptr;
+
+        // First part:
+        // Filter out predicate DoFs that are constrained to DoFs in the
+        // halo layer, that does not satisfy predicate.
+
+        // TODO: forgetting about hp, one can optimize this by only going
+        // through DoFs that are in the inner layer of the locally owned
+        // predicate. As of now, bulk of constrained DoFs in predicate_dofs
+        // are cosntrained to DoFs within the same set.
+        for (std::set<types::global_dof_index>::iterator it = predicate_dofs.begin();
+             it != predicate_dofs.end();)
+          {
+            bool erase = false;
+            line_ptr = cm.get_constraint_entries(*it);
+            // predicate DoF is constrained, see if it's constarined to
+            // a DoF in the halo layer. If so -- remove it.
+            if (line_ptr!=NULL)
```

Comment:
the two lines here can probably be written more easily as `if (cm.is_constrained(*it))`.